src/google/protobuf/compiler/ruby/rbs_generator.cc

```diff
@@ -0,0 +1,806 @@
+// Protocol Buffers - Google's data interchange format
+// Copyright 2008 Google Inc.  All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file or at
+// https://developers.google.com/open-source/licenses/bsd
+
+#include "google/protobuf/compiler/ruby/rbs_generator.h"
+
+#include "absl/strings/str_split.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_replace.h"
+#include "google/protobuf/io/printer.h"
+#include "google/protobuf/compiler/ruby/ruby_generator.h"
+
+namespace google {
+namespace protobuf {
+namespace compiler {
+namespace ruby {
+
+std::string GetRBSOutputFilename(absl::string_view proto_file) {
+  return absl::StrCat(GetRequireName(proto_file), ".rbs");
+}
+
+std::string WrapUnion(const std::string& type) {
+  // This is known not to work as intended for ::Array[::Integer | ::Float].
+  // Maybe we need an AST
+  if (type.find('|') != std::string::npos) {
+    return absl::StrCat("(", type, ")");
+  } else {
+    return absl::StrCat(type);
+  }
+}
+
+std::string MakeOptionalType(const std::string& type) {
+  if (type.back() == '?') {
+    return type;
+  } else {
+    return absl::StrCat(WrapUnion(type), "?");
+  }
+}
+
+void PrintComment(const std::string& comment, io::Printer* printer) {
+  if (comment.empty()) {
+    return;
+  }
+  std::vector<std::string> lines = absl::StrSplit(comment, '\n', absl::SkipEmpty());
+  if (!lines.empty() && lines.back().empty()) {
+    lines.pop_back();
+  }
+  for (const std::string& line : lines) {
+    // Spaces after `#` are usually included in the comment itself
+    printer->Print("#$line$\n", "line", line);
+  }
+}
+
+void PrintDeprecationComment(bool deprecated, io::Printer* printer) {
+  if (deprecated) {
+    printer->Print("# @deprecated\n");
+  }
+}
+
+void InsertMidLine(bool &initial, io::Printer* printer) {
+  if (initial) {
+    initial = false;
+  } else {
+    printer->Print("\n");
+  }
+}
+
+std::string UnionSeparator(bool &initial) {
+  if (initial) {
+    initial = false;
+    return "  ";
+  } else {
+    return "| ";
+  }
+}
+
+std::string ModulePath(const FileDescriptor *file) {
+  // Mirror implementation of GeneratePackageModules
+  bool need_change_to_module = true;
+  std::string package_name;
+  if (file->options().has_ruby_package()) {
+    package_name = file->options().ruby_package();
+
+    if (package_name.find("::") != std::string::npos) {
+      need_change_to_module = false;
+    }
+  } else {
+    package_name = file->package();
```

Comment:
The line is getting flagged with an error on some builds and not others. Here's one from a [failing GCC-based build](https://github.com/protocolbuffers/protobuf/actions/runs/14271858302/job/40006682513?pr=15633#step:4:4025).

Elsewhere, I see similar lines constructed as:
`package_name = std::string(file->package())`

Reply:
@JasonLunn done.